Adjusted mastodon-tl.el to open user profiles in emacs.

alexjgriffith opened the pull request:
This PR preps for opening user profiles within emacs.

## DESCRIPTION
Rather than opening a users profile in a browser, mastodon.el now searches for the user handle and returns an account or a nil. The account can be fed directly into `mastodon-profile--make-author-buffer` once PR #186 is merged.

alexjgriffith committed:
Adjusted mastodon-tl.el to open user profiles in emacs.

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -244,9 +244,10 @@ Optionally start from POS."
                  'mouse-face 'highlight
 		 ;; TODO: Replace url browsing with native profile viewing
-		 'mastodon-tab-stop 'shr-url
+		 'mastodon-tab-stop 'handle
```
`user-handle` or some thing like that?  "handle" is awfully generic.

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -244,9 +244,10 @@ Optionally start from POS."
 		 'shr-url profile-url
-		 'keymap mastodon-tl--shr-map-replacement
+		 'keymap mastodon-tl--link-keymap
+                 'mastodon-handle (concat "@" handle)
```
Here we actually have the account info, why not store that in the property (or some other property)?
Seems wasteful (resource wasting for the server and time wasting for the user) to go back to the server to look up information we already have.

alexjgriffith replied:
Good call. Since the account is in the toot, we can check if the handle has 'toot-json before doing a search and extract the account from there.

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -415,10 +416,10 @@ links in the text."
+           (setq mastodon-tab-stop-type 'handle
+                 keymap mastodon-tl--link-keymap
+                 help-echo (concat "Browse user profile of " maybe-userhandle)
+                 extra-properties (list 'mastodon-handle maybe-userhandle)))
```
We can also look up the id in the toot's `mention` field.
With that we still have to go to the server to look up the full account details but getting details for a specific account (via [`GET /api/v1/accounts/:id`](https://github.com/tootsuite/documentation/blob/master/Using-the-API/API.md#accounts)) should be much lighter on the server.

alexjgriffith replied:
That would work for byline handles, but do inline handles have id's in their meta info? For the time being we will look up the account id in 'toot-json, if it doesn't exist we can do a search.

alexjgriffith replied:
I see, the ids are in the mentions, will save a lot of search calls.

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -512,13 +513,33 @@ LINK-TYPE is the type of link to produce."
      'keymap mastodon-tl--link-keymap
      'help-echo help-text)))
 
+(defun mastodon-tl--handle-to-account (handle)
```
Name this something like "search account by handle"?  And place this in `mastodon-profile.el`?

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -512,13 +513,33 @@ LINK-TYPE is the type of link to produce."
+                   handle))
+         (matching-account
+          (remove-if-not
+           (lambda(x) (string= (cdr (assoc 'acct x)) handle))
```
Do we need to worry about case sensitivity here?

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -513,13 +514,33 @@ LINK-TYPE is the type of link to produce."
            (mastodon-tl--show-tag-timeline (get-text-property position 'mastodon-tag)))
+          ((eq link-type 'handle)
+           ;; replace message with mastodon-profile--make-author-buffer
+           (message "its a handle: %s"
```
In my cursory testing this seems to work.
Get rid of the `message` and instead actually show the profile.

hdurer commented on test/mastodon-tl-tests.el:
```diff
@@ -266,7 +266,7 @@ a string or a numeric."
   ------------"))
 	(should (eq (get-text-property handle-location 'mastodon-tab-stop byline)
-                'shr-url))
+                'handle))
```
We should probably also assert that there is the expected property with the account data (handle name at the moment) at the point.

alexjgriffith committed:
Fixes for opening user profiles interactively.

- new updates now appear below the profile note
- shr property has been replaced with user-handle for user handles
- Rather than doing a search for userids based on their handles we now either provide the account strucure from the toot or the account id from the mentions. We still preform a search if both of those are missing.
- mastodon-profile--make-author-buffer hooked in
- mastodon-tl--handle-to-account has been moved to profiles and now is called mastodon-profile-serach-account-by-handle
- mastodon-tl--byline-regular test expanded
